pico/services/src/main/java/io/helidon/pico/services/DefaultPicoServices.java

```diff
@@ -0,0 +1,528 @@
+                if (!fullyActivationLog.isEmpty()) {
+                    LinkedHashSet<ServiceProvider<?>> serviceProviderActivations = new LinkedHashSet<>();
+
+                    Collections.reverse(fullyActivationLog);
+                    fullyActivationLog.stream()
+                            .map(ActivationLogEntry::serviceProvider)
+                            .filter(Optional::isPresent)
+                            .map(Optional::get)
+                            .forEach(serviceProviderActivations::add);
+
+                    // prepare for the shutdown log event sequence
+                    log.toQuery().ifPresent(it -> it.reset(false));
+
+                    // shutdown using the reverse chronological ordering in the log for starters
+                    doFinalShutdown(serviceProviderActivations);
+                }
+            }
+
+            // next get all services that are beyond INIT state, and sort by runlevel order, and shut those down also
+            List<ServiceProvider<?>> serviceProviders = services.lookupAll(DefaultServiceInfoCriteria.builder().build(), false);
+            serviceProviders = serviceProviders.stream()
+                    .filter(sp -> sp.currentActivationPhase().eligibleForDeactivation())
+                    .collect(Collectors.toList());
+            serviceProviders.sort((o1, o2) -> {
+                int runLevel1 = o1.serviceInfo().realizedRunLevel();
+                int runLevel2 = o2.serviceInfo().realizedRunLevel();
+                return Integer.compare(runLevel1, runLevel2);
+            });
+            doFinalShutdown(serviceProviders);
+
+            // finally, clear everything
+            reset(false);
+
+            return map;
+        }
+
+        private void doFinalShutdown(Collection<ServiceProvider<?>> serviceProviders) {
+            for (ServiceProvider<?> csp : serviceProviders) {
+                Phase startingActivationPhase = csp.currentActivationPhase();
+                ActivationResult result;
+                try {
+                    result = injector.deactivate(csp, opts);
+                } catch (Throwable t) {
+                    errorLog("error during shutdown", t);
+                    result = DefaultActivationResult.builder()
+                            .serviceProvider(csp)
+                            .startingActivationPhase(startingActivationPhase)
+                            .targetActivationPhase(Phase.DESTROYED)
+                            .finishingActivationPhase(csp.currentActivationPhase())
+                            .finishingStatus(ActivationStatus.FAILURE)
+                            .error(t)
+                            .build();
+                }
+                map.put(csp.serviceInfo().serviceTypeName(), result);
+            }
+        }
+    }
+
+    @Override
+    // note that this is typically only called during testing, and also in the pico-maven-plugin
+    public synchronized boolean reset(boolean deep) {
```

Comment:
Do not use `synchronized`. Use a lock or a semaphore instead if you need to

Reply:
pushed

Reply:
i suggest we file this as a follow-up. Someone would need to be doing either `shutdown()` or `reset()` in a different thread literally at the same time as doing `startup()`, which seems extremely improbable in the near to medium term.

Reply:
https://github.com/helidon-io/helidon/issues/6498